src/hooks/useViewport/useViewport.ts

```diff
@@ -0,0 +1,31 @@
+import { useLayoutEffect, useState } from 'react';
+import throttle from 'lodash.throttle';
+
+interface IUseViewportOptions {
+  timeout?: number;
+}
+
+function readViewport() {
+  const width = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
+  const height = Math.max(document.documentElement.clientHeight, window.innerHeight || 0);
+
+  return {
+    height,
+    width,
+  };
+}
+
+export function useViewport({ timeout = 300 }: IUseViewportOptions = {}) {
+  const [size, setSize] = useState(readViewport());
```

Comment:
This might be quite heavy if someone start playing with the window size ?

Reply:
Exact. We are generating an eventListener every time some component calls the useViewport hook. Maybe, we can improve this using a context.

Reply:
@Blimeys won't be a problem as the event listener is throttled 💅🏻

@dfvalero you're right on the fact that a listener is added to `window` **every time** a mounted component uses it 🤧 . However, I was initially drawn away from forcing context on this ( _seemed a bit too much..._ 💄 ) maybe we could enable it via an option?
```jsx
/**
 * When supplying this option the hook will throw unless there's
 * a <ViewportProvider /> above on the tree. In this mode the hook
 * will "connect" to the context rather than attach its ow listener
 */
function Button() {
  const {width} = useViewport({ context: true })
  // ...
}

/**
 * Reads viewports sizes from its own listener (as it's today) 
 */ 
function Table() {
  const {width} = useViewport()
}
```

In this way, you won't have an extra context in your app unless you have a use case for it having multiple components on a single screen using this hook.

Reply:
What about this:
- We create the Context
- if we call useViewport in a component that is a child of the Context Provider, we use the context.
- if we call useVieport in a component that is **not** a child of the Context Provider, we use a regular useEffect solution

too complicated?

Reply:
@dfvalero sounds good to me 💯 , to clarify this is how it ends:
```jsx
import { useViewport, ViewportContext } from '@zopauk/react-components'

const Button = () => {
  const {width} = useViewport()
  // ...
}

const App1 = () => (
  <ViewportContext.Provider>
    <Button /> {/* <Button /> will use context given there's a provider above */}
  </ViewportContext.Provider>
) 

const App2 = () => (
  <>
    <Button /> {/* <Button /> will uses listener given ther's no context above */}
  </>
) 
```

Reply:
beeautiiiiiful <3